device/cuda/src/clusterization/clusterization_algorithm.cpp

```diff
@@ -0,0 +1,143 @@
+    traccc::cuda::clusters_sum(cells_view, sparse_ccl_indices, *total_clusters,
+                               cluster_prefix_sum, clusters_per_module);
+
+    // Vector of the exact cluster sizes, will be filled in cluster counting
+    vecmem::data::vector_buffer<unsigned int> cluster_sizes_buffer(
+        *total_clusters, m_mr.get());
+    copy.setup(cluster_sizes_buffer);
+    copy.memset(cluster_sizes_buffer, 0);
+    traccc::cuda::cluster_counting(sparse_ccl_indices, cluster_sizes_buffer,
+                                   cluster_prefix_sum,
+                                   vecmem::get_data(cells_prefix_sum));
+
+    std::vector<unsigned int> cluster_sizes;
+    copy(cluster_sizes_buffer, cluster_sizes);
+
+    // Cluster container buffer for the clusters and headers (cluster ids)
+    cluster_container_types::buffer clusters_buffer{
+        {*total_clusters, m_mr.get()},
+        {std::vector<std::size_t>(*total_clusters, 0),
+         std::vector<std::size_t>(cluster_sizes.begin(), cluster_sizes.end()),
+         m_mr.get()}};
+
+    copy.setup(clusters_buffer.headers);
+    copy.setup(clusters_buffer.items);
+
+    // Component connection kernel
+    traccc::cuda::component_connection(clusters_buffer, cells_view,
+                                       sparse_ccl_indices, cluster_prefix_sum,
+                                       vecmem::get_data(cells_prefix_sum));
+
+    // Copy the sizes of clusters per each module to the std vector for
+    // measurement buffer initialization
+    std::vector<std::size_t> clusters_per_module_host;
+    copy(clusters_per_module, clusters_per_module_host);
+
+    // Resizable buffer for the measurements
+    measurement_container_types::buffer measurements_buffer{
+        {num_modules, m_mr.get()},
+        {std::vector<std::size_t>(num_modules, 0), clusters_per_module_host,
+         m_mr.get()}};
+    copy.setup(measurements_buffer.headers);
+    copy.setup(measurements_buffer.items);
+    // Measurement creation kernel
+    traccc::cuda::measurement_creation(measurements_buffer, clusters_buffer,
+                                       cells_view);
+
+    // Spacepoint container buffer to fill in spacepoint formation
+    spacepoint_container_types::buffer spacepoints_buffer{
+        {num_modules, m_mr.get()},
+        {std::vector<std::size_t>(num_modules, 0), clusters_per_module_host,
+         m_mr.get()}};
+    copy.setup(spacepoints_buffer.headers);
+    copy.setup(spacepoints_buffer.items);
+
+    // Get the prefix sum of the measurements.
+    const device::prefix_sum_t measurements_prefix_sum = device::get_prefix_sum(
+        copy.get_sizes(measurements_buffer.items), m_mr.get());
+    // Spacepoint formation kernel
+    traccc::cuda::spacepoint_formation(
+        spacepoints_buffer, measurements_buffer,
+        vecmem::get_data(measurements_prefix_sum));
+
+    spacepoint_container_types::host spacepoints_cuda(num_modules, &m_mr.get());
```

Comment:
It's where the speed gets slower. Generating host container with vecmem jagged vector is pretty expensive. 
I think you need to get buffer object as an output...
@konradkusiak97 What is the actual plan for object type in algorithm I/O? SYCL is using the buffer output from clusterization algorithm. I know that the overhead from host_container can be minimized once we start using the caching allocator but I still like using view type object for algorithm I/O

Reply:
I thought about this and went ahead and changed all instances that expect a host container to views in the seeding algorithm, consequently few more places require similar changes. So i did not continue it with it.
Maybe we can discuss this a little bit tomorrow?

Reply:
Sure